src/main/java/org/broadinstitute/hellbender/tools/spark/validation/CompareBaseQualities.java

```diff
@@ -0,0 +1,235 @@
+        ReadsSparkSource readsSource2 = new ReadsSparkSource(ctx);
+        JavaRDD<GATKRead> secondReads = readsSource2.getParallelReads(input2, getIntervals());
+
+        long firstBamSize = firstReads.count();
+        long secondBamSize = secondReads.count();
+        if (firstBamSize != secondBamSize) {
+            throw new UserException("input bams have different numbers of total reads: "
+                    + firstBamSize + "," + secondBamSize);
+        }
+        logger.info("Processing bams with " + firstBamSize + " reads");
+
+        // To compare bases, we created two PairRDDs keyed by read name, strand, etc. where the value is the bases.
+        // We cogroup the two RDDs, verify there is only one read per key per bam, and then diff the bases
+        // position by position for the two reads and put the results in CompareMatrix.
+        JavaPairRDD<String, Quals> firstQuals = firstReads.mapToPair(read -> {
+            // For both BAMs, we clear the attributes because want to key by common unique per-read properties, start, contig, etc.
+            read.clearAttributes();
+            return new Tuple2<>(BaseQualsUtils.getReadKey(read), new Quals(read.getBaseQualities()));
+        });
+        JavaPairRDD<String, Quals> secondQuals = secondReads.mapToPair(read -> {
+            read.clearAttributes();
+            return new Tuple2<>(BaseQualsUtils.getReadKey(read), new Quals(read.getBaseQualities()));
+        });
+
+        JavaPairRDD<String, Tuple2<Iterable<Quals>, Iterable<Quals>>> cogroup = firstQuals.cogroup(secondQuals);
+        CompareMatrix finalMatrix = cogroup.map(v1 -> {
+            List<Quals> lFirstQuals = Lists.newArrayList(v1._2()._1());
+            List<Quals> lSecondQuals = Lists.newArrayList(v1._2()._2());
+            CompareMatrix compareMatrix = new CompareMatrix();
+            if (lFirstQuals.size() != 1) {
+                throw new GATKException("expected only one read per key in first bam: " + lFirstQuals.size());
+            }
+            if (lSecondQuals.size() != 1) {
+                throw new GATKException("expected only one read per key in second bam: " + lSecondQuals.size());
+            }
+
+            compareMatrix.add(lFirstQuals.get(0).quals, lSecondQuals.get(0).quals);
+            return compareMatrix;
+        }).treeAggregate(new CompareMatrix(),
+                (Function2<CompareMatrix, CompareMatrix, CompareMatrix>) CompareMatrix::addMatrix,
+                (Function2<CompareMatrix, CompareMatrix, CompareMatrix>) CompareMatrix::addMatrix);
+
+
+        if (printSummary) {
+            finalMatrix.printSummary();
+        }
+
+        if (printMatrix) {
+            finalMatrix.print();
+        }
+
+        if (crashOnDiff && finalMatrix.hasNonDiagonalElements()) {
+            throw new UserException("Quality scores from the two BAMs do not match");
+        }
+    }
+
+}
+
+class BaseQualsUtils implements Serializable {
```

Comment:
only 1 top-level class per file. 
Put this in another file or make it a `static final class` inside `CompareBaseQualities`